File: 1160.rb
# @param {String} word
# @param {String} chars
# @return {Boolean}
def formable?(word, chars)
  array = chars.chars
  word.each_char do |c|
    index = array.index(c)
    return false unless index
    array.delete_at(index)
  end
  true
end

# @param {String[]} words
# @param {String} chars
# @return {Integer}
def count_characters(words, chars)
    result = 0
    words.each do |word|
      result += word.length if formable?(word, chars)
    end
    result
end

# p count_characters(["cat","bt","hat","tree"], "atach")
# p count_characters(["hello","world","leetcode"], "welldonehoneyr")

File: 1160.ts
/**
 * @param {string} word
 * @param {object} hash
 * @return {boolean}
 */
function formable(word: string, hash: object): boolean {
  for (let c of word) {
    hash[c] = (hash[c] || 0) - 1
    if (hash[c] < 0) {
      return false
    }
  }
  return true
}

/**
 * @param {string[]} words
 * @param {string} chars
 * @return {number}
 */
function countCharacters(words: string[], chars: string): number {
  const hash = {}
  for (let c of chars) {
    hash[c] = (hash[c] || 0) + 1
  }
  return words.reduce((acc, cur) => (acc += formable(cur, Object.assign({}, hash)) ? cur.length : 0), 0)
}

// console.log(countCharacters(["cat","bt","hat","tree"], "atach"));
// console.log(countCharacters(["hello","world","leetcode"], "welldonehoneyr"));
